File: DemonHunterDemonic.lua
local OVALE, Ovale = ...
local OvaleDemonHunterDemonic = Ovale:NewModule("OvaleDemonHunterDemonic", "AceEvent-3.0")
Ovale.OvaleDemonHunterDemonic = OvaleDemonHunterDemonic

-- Forward declarations for module dependencies.
local OvaleDebug = nil
local OvaleAura = nil

local API_GetSpecialization = GetSpecialization
local API_GetSpecializationInfo = GetSpecializationInfo
local API_GetTime = GetTime
local API_GetTalentInfoByID = GetTalentInfoByID
local INFINITY = math.huge

local HAVOC_DEMONIC_TALENT_ID = 22547
local HAVOC_SPEC_ID = 577
local HAVOC_EYE_BEAM_SPELL_ID = 198013
local HAVOC_META_BUFF_ID = 162264

local HIDDEN_BUFF_ID = -HAVOC_DEMONIC_TALENT_ID
local HIDDEN_BUFF_DURATION = INFINITY
local HIDDEN_BUFF_EXTENDED_BY_DEMONIC = "Extended by Demonic"


function OvaleDemonHunterDemonic:OnInitialize()
	-- Resolve module dependencies.
	OvaleAura = Ovale.OvaleAura
	OvaleDebug = Ovale.OvaleDebug
	
	-- Register for debugging messages.
	OvaleDebug:RegisterDebugging(OvaleDemonHunterDemonic)
end

function OvaleDemonHunterDemonic:OnEnable()
	self.playerGUID = nil
	self.isDemonHunter = Ovale.playerClass == "DEMONHUNTER" and true or false
	self.isHavoc = false
	self.hasDemonic = false
	
	if self.isDemonHunter then
		self:Debug("playerGUID: (%s)", Ovale.playerGUID)
		self.playerGUID = Ovale.playerGUID
		self:RegisterMessage("Ovale_TalentsChanged")
	end
end

function OvaleDemonHunterDemonic:OnDisable()
	self:UnregisterMessage("COMBAT_LOG_EVENT_UNFILTERED")
end

function OvaleDemonHunterDemonic:Ovale_TalentsChanged(event)
	-- are we havoc?
	self.isHavoc = self.isDemonHunter and API_GetSpecializationInfo(API_GetSpecialization()) == HAVOC_SPEC_ID and true or false
	-- do we have Demonic talented.
	self.hasDemonic = self.isHavoc and select(10, API_GetTalentInfoByID(HAVOC_DEMONIC_TALENT_ID, HAVOC_SPEC_ID)) and true or false
	
	-- register events
	if self.isHavoc and self.hasDemonic then
		self:Debug("We are a havoc DH with Demonic.")
		self:RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED")
	else
		if not self.isHavoc then
			self:Debug("We are not a havoc DH.")
		elseif not self.hasDemonic then
			self:Debug("We don't have the Demonic talent.")
		end
		self:DropAura()
		self:UnregisterMessage("COMBAT_LOG_EVENT_UNFILTERED")
	end
end

function OvaleDemonHunterDemonic:COMBAT_LOG_EVENT_UNFILTERED(event, timestamp, cleuEvent, hideCaster, sourceGUID, sourceName, sourceFlags, sourceRaidFlags, destGUID, destName, destFlags, destRaidFlags, ...)
	local arg12, arg13, arg14, arg15, arg16, arg17, arg18, arg19, arg20, arg21, arg22, arg23, arg24, arg25 = ...
	if sourceGUID == self.playerGUID and cleuEvent == "SPELL_CAST_SUCCESS" then
		local spellId, spellName = arg12, arg13
		if HAVOC_EYE_BEAM_SPELL_ID == spellId then
			self:Debug("Spell %d (%s) has successfully been cast. Gaining Aura (only during meta).", spellId, spellName)
			self:GainAura()
		end
	end
	if sourceGUID == self.playerGUID and cleuEvent == "SPELL_AURA_REMOVED" then
		local spellId, spellName = arg12, arg13
		if HAVOC_META_BUFF_ID == spellId then
			self:Debug("Aura %d (%s) is removed. Dropping Aura.", spellId, spellName)
			self:DropAura()
		end
	end
end

function OvaleDemonHunterDemonic:GainAura()
	-- Add the "extended by demonic buff"
	local now = API_GetTime()	
	local aura_meta = OvaleAura:GetAura("player", HAVOC_META_BUFF_ID, "HELPFUL", true)
	if OvaleAura:IsActiveAura(aura_meta, now) then
		self:Debug("Adding '%s' (%d) buff to player %s.", HIDDEN_BUFF_EXTENDED_BY_DEMONIC, HIDDEN_BUFF_ID, self.playerGUID)
		local duration = HIDDEN_BUFF_DURATION
		local ending = now + HIDDEN_BUFF_DURATION
		OvaleAura:GainedAuraOnGUID(self.playerGUID, now, HIDDEN_BUFF_ID, self.playerGUID, "HELPFUL", nil, nil, 1, nil, duration, ending, nil, HIDDEN_BUFF_EXTENDED_BY_DEMONIC, nil, nil, nil)
	else
		self:Debug("Aura 'Metamorphosis' (%d) is not present.", HAVOC_META_BUFF_ID)
	end
	
end

function OvaleDemonHunterDemonic:DropAura()
	-- drop the aura
	local now = API_GetTime()	
	self:Debug("Removing '%s' (%d) buff on player %s.", HIDDEN_BUFF_EXTENDED_BY_DEMONIC, HIDDEN_BUFF_ID, self.playerGUID)
	OvaleAura:LostAuraOnGUID(self.playerGUID, now, HIDDEN_BUFF_ID, self.playerGUID)
end

File: src/DemonHunterDemonic.ts
import { Ovale } from "./Ovale";
import { OvaleDebug } from "./Debug";
import { OvaleAura } from "./Aura";
import aceEvent from "@wowts/ace_event-3.0";
import { GetSpecialization, GetSpecializationInfo, GetTime, GetTalentInfoByID, CombatLogGetCurrentEventInfo } from "@wowts/wow-mock";
import { huge } from "@wowts/math";
import { select } from "@wowts/lua";

let OvaleDemonHunterDemonicBase = OvaleDebug.RegisterDebugging(Ovale.NewModule("OvaleDemonHunterDemonic", aceEvent));
export let OvaleDemonHunterDemonic: OvaleDemonHunterDemonicClass;
let INFINITY = huge;
let HAVOC_DEMONIC_TALENT_ID = 22547;
let HAVOC_SPEC_ID = 577;
let HAVOC_EYE_BEAM_SPELL_ID = 198013;
let HAVOC_META_BUFF_ID = 162264;
let HIDDEN_BUFF_ID = -HAVOC_DEMONIC_TALENT_ID;
let HIDDEN_BUFF_DURATION = INFINITY;
let HIDDEN_BUFF_EXTENDED_BY_DEMONIC = "Extended by Demonic";
class OvaleDemonHunterDemonicClass extends OvaleDemonHunterDemonicBase {
    playerGUID:string;
    isDemonHunter:boolean;
    isHavoc:boolean;
    hasDemonic: boolean;

    OnInitialize() {
        this.playerGUID = undefined;
        this.isDemonHunter = Ovale.playerClass == "DEMONHUNTER" && true || false;
        this.isHavoc = false;
        this.hasDemonic = false;
        if (this.isDemonHunter) {
            this.Debug("playerGUID: (%s)", Ovale.playerGUID);
            this.playerGUID = Ovale.playerGUID;
            this.RegisterMessage("Ovale_TalentsChanged");
        }
    }
    OnDisable() {
        this.UnregisterMessage("COMBAT_LOG_EVENT_UNFILTERED");
    }
    Ovale_TalentsChanged(event) {
        this.isHavoc = this.isDemonHunter && GetSpecializationInfo(GetSpecialization()) == HAVOC_SPEC_ID && true || false;
        this.hasDemonic = this.isHavoc && select(10, GetTalentInfoByID(HAVOC_DEMONIC_TALENT_ID, HAVOC_SPEC_ID)) && true || false;
        if (this.isHavoc && this.hasDemonic) {
            this.Debug("We are a havoc DH with Demonic.");
            this.RegisterEvent("COMBAT_LOG_EVENT_UNFILTERED");
        } else {
            if (!this.isHavoc) {
                this.Debug("We are not a havoc DH.");
            } else if (!this.hasDemonic) {
                this.Debug("We don't have the Demonic talent.");
            }
            this.DropAura();
            this.UnregisterMessage("COMBAT_LOG_EVENT_UNFILTERED");
        }
    }
    COMBAT_LOG_EVENT_UNFILTERED(event: string, ...__args: any[]) {
        let [, cleuEvent, , sourceGUID, , , , , , , , arg12, arg13] = CombatLogGetCurrentEventInfo();
        if (sourceGUID == this.playerGUID && cleuEvent == "SPELL_CAST_SUCCESS") {
            let [spellId, spellName] = [arg12, arg13];
            if (HAVOC_EYE_BEAM_SPELL_ID == spellId) {
                this.Debug("Spell %d (%s) has successfully been cast. Gaining Aura (only during meta).", spellId, spellName);
                this.GainAura();
            }
        }
        if (sourceGUID == this.playerGUID && cleuEvent == "SPELL_AURA_REMOVED") {
            let [spellId, spellName] = [arg12, arg13];
            if (HAVOC_META_BUFF_ID == spellId) {
                this.Debug("Aura %d (%s) is removed. Dropping Aura.", spellId, spellName);
                this.DropAura();
            }
        }
    }
    GainAura() {
        let now = GetTime();
        let aura_meta = OvaleAura.GetAura("player", HAVOC_META_BUFF_ID, undefined, "HELPFUL", true);
        if (OvaleAura.IsActiveAura(aura_meta, now)) {
            this.Debug("Adding '%s' (%d) buff to player %s.", HIDDEN_BUFF_EXTENDED_BY_DEMONIC, HIDDEN_BUFF_ID, this.playerGUID);
            let duration = HIDDEN_BUFF_DURATION;
            let ending = now + HIDDEN_BUFF_DURATION;
            OvaleAura.GainedAuraOnGUID(this.playerGUID, now, HIDDEN_BUFF_ID, this.playerGUID, "HELPFUL", undefined, undefined, 1, undefined, duration, ending, undefined, HIDDEN_BUFF_EXTENDED_BY_DEMONIC, undefined, undefined, undefined);
        } else {
            this.Debug("Aura 'Metamorphosis' (%d) is not present.", HAVOC_META_BUFF_ID);
        }
    }
    DropAura() {
        let now = GetTime();
        this.Debug("Removing '%s' (%d) buff on player %s.", HIDDEN_BUFF_EXTENDED_BY_DEMONIC, HIDDEN_BUFF_ID, this.playerGUID);
        OvaleAura.LostAuraOnGUID(this.playerGUID, now, HIDDEN_BUFF_ID, this.playerGUID);
    }
}

OvaleDemonHunterDemonic = new OvaleDemonHunterDemonicClass();